### ISSUE TITLE
Remote issues with clipboard and NativeImage

### ISSUE THREAD

**Issue by bahmutov** [electron/electron#8151]
* Electron version: 1.4.10
* Operating system: MacOS

### Expected behavior

Captured image from webview is copied into clipboard

### Actual behavior

Crashes with error message 

```
Uncaught TypeError: Error processing argument at index 0, conversion failure from [object Object]
```

### How to reproduce

I am capturing a region of a webview object and then saving as an image (works pretty ok, although sometimes the image is empty), then I am trying to copy the NativeImage into clipboard, but it crashes

```js
webview.capturePage(rect, image => {
              const quality = 90
              const filename = 'image.jpg'
              const fullPath = path.join(__dirname, filename)
              fs.writeFile(fullPath, image.toJPEG(quality), (err) => {                
                 // TODO: retry save if the image is 0 size
                clipboard.writeImage(image)
              })
})
```

I can see the image on the disk, but the clipboard has never received it, always crashing

**Comment by kevinsawicki** [comment]
> Uncaught TypeError: Error processing argument at index 0, conversion failure from [object Object]

The argument to `writeImage` needs to be a `NativeImage` or `String` file path, `image.toJPEG()` returns a `Buffer`.

Use `nativeImage.fromBuffer(image.toJPEG(quality))` instead.

http://electron.atom.io/docs/api/native-image/

**Comment by bahmutov** [comment]
Yes, so in the code example, `webview.capturePage` returns NativeImage, which I am passing directly into `clipboard.writeImage(image)`. Conversion to JPEG is only done to save the image to disk to make sure the file is valid.

**Comment by kevinsawicki** [comment]
> Conversion to JPEG is only done to save the image to disk to make sure the file is valid.

Whoops, sorry, misread that.

**Comment by kevinsawicki** [comment]
You can wrap it in a `nativeImage.createFromBuffer(image.getBitmap())` to workaround this.

This issue is happening because `webview.capturePage` is actually a remote method doing IPC so the value it calls back with is a remote `nativeImage` object. It does not seem like it is currently supported to pass these back into subsequent API calls.

**Comment by bahmutov** [comment]
so far no luck, does not seem the image is placed into the clipboard. Is there a good tool to inspect the clipboard for Mac?

**Comment by kevinsawicki** [comment]
> Is there a good tool to inspect the clipboard for Mac?

I usually just launch `TextEdit` since it supports pasting images.

Can you try `clipboard.writeImage(nativeImage.createFromBuffer(image.toJPEG(1000)))` instead?

**Comment by kevinsawicki** [comment]
`clipboard.writeImage(nativeImage.createFromBuffer(image.toPNG()))` works for me too.

**Comment by bahmutov** [comment]
Cool, both the above work arounds work, and the image can be pasted into Chrome email clients for example. Thank you very much for looking at this.

**Comment by kevinsawicki** [comment]
> Cool, both the above work arounds work

Great to hear, you can avoid these workarounds if you run this from the main process.

### In Main Processs

```js
ipcMain.on('capturePageToClipboard', (event, contentsId, rect) => {
  webContents.fromId(contentsId).capturePage(rect, (image) => {
    clipboard.writeImage(image)
  })
})
```

### In Render Process

```js
const capturePageToClipboard = (rect) => {
  ipcRenderer.send('capturePageToClipboard', webview.getWebContents().id, rect)
} 
```

**Comment by kevinsawicki** [comment]
This can also occur the other way (#8665), you cannot currently pass a renderer process `NativeImage` to `remote.clipboard.writeImage`. This affects Linux since the clipboard module in the renderer process uses `remote`.

**Comment by momocow** [comment]
I am also looking forward to see this problem gets fixed.

Should this problem (or the workarounds) be mentioned in the [<webview>.capturePage()](https://electron.atom.io/docs/api/webview-tag/#webviewcapturepagerect-callback)?
Since this issue is filed almost 11 months ago and has not been fixed in electron 1.7.9, which is the electron version which I am depending on and I got the same error message too.

**Comment by DuBistKomisch** [comment]
Seeing a similar issue, more like https://github.com/electron/electron/issues/8665 as it's passing a NativeImage on Linux.

From:

```
    let image = nativeImage.createFromDataURL(dataURL);
    clipboard.writeImage(image, 'png');
```

To:

```
    ipcRenderer.send('saveDataURLToClipboard', dataURL);
```

```
ipcMain.on('saveDataURLToClipboard', (event, dataURL) => {
  let image = electron.nativeImage.createFromDataURL(dataURL);
  clipboard.writeImage(image, 'png');
});
```

Works I guess, hopefully not too much data to be sent via IPC.

**Comment by sofianguy** [comment]
Thank you for taking the time to report this issue and helping to make Electron better.

The version of Electron you reported this on has been superseded by newer releases. 

If you're still experiencing this issue in Electron v4.2.x or later, please add a comment specifying the version you're testing with and any other new information that a maintainer trying to reproduce the issue should know.

I'm setting the `blocked/need-info` label for the above reasons. This issue will be closed 7 days from now if there is no response.

Thanks in advance! Your help is appreciated.

**Comment by DuBistKomisch** [comment]
This is still an issue on 5.0.9:

```
Uncaught Error: Could not call remote function 'anonymous'. Check that the function signature is correct. Underlying error: Error processing argument at index 0, conversion failure from #<Object>
```
at `renderer/api/remote.js:222`

**Comment by miniak** [comment]
Will be possible in Electron 6+ after #20825 is merged and backported
